randovania/game_description/world/dock_lock_node.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+
+import dataclasses
+from typing import Iterator
+
+from randovania.game_description.requirements import Requirement
+from randovania.game_description.resources.resource_info import ResourceInfo, ResourceGain
+from randovania.game_description.world.dock_node import DockNode
+from randovania.game_description.world.node import NodeContext, Node
+from randovania.game_description.world.node_identifier import NodeIdentifier
+from randovania.game_description.world.resource_node import ResourceNode
+
+
+@dataclasses.dataclass(frozen=True)
+class DockLockNode(ResourceNode):
+    dock_identifier: NodeIdentifier
+
+    @classmethod
+    def create_from_dock(cls, dock: DockNode, dock_identifier: NodeIdentifier, index: int) -> DockLockNode:
+        lock_identifier = dock.get_lock_node_identifier_from_identifier(dock_identifier)
+        return DockLockNode(
+            name=lock_identifier.node_name,
+            heal=False,
+            location=None,
+            description="",
+            extra={},
+            index=index,
+            dock_identifier=dock_identifier,
+        )
+
+    def __repr__(self):
+        return "DockLockNode({!r} -> {})".format(self.name, self.dock_identifier)
+
+    def resource(self, context: NodeContext) -> ResourceInfo:
+        return self.dock_identifier
+
+    def can_collect(self, context: NodeContext) -> bool:
+        dock = self._get_dock(context)
+
+        front_weak = dock.get_front_weakness(context)
+        if not context.has_resource(self.dock_identifier):
+            if front_weak.lock is not None:
+                return True
+
+        dock_target = dock.get_target_identifier(context)
+        if not context.has_resource(dock_target):
+            if front_weak.can_unlock_from_back(dock.get_back_weakness(context)):
+                return True
+
+        return False
+
+    def is_collected(self, context: NodeContext) -> bool:
+        dock = self._get_dock(context)
+
+        front_weak = dock.get_front_weakness(context)
+        if front_weak.lock is not None:
+            if not context.has_resource(self.dock_identifier):
+                return False
+
+        dock_target = dock.get_target_identifier(context)
+        if front_weak.can_unlock_from_back(dock.get_back_weakness(context)):
+            if not context.has_resource(dock_target):
+                return False
+
+        return True
+
```

Comment:
couldn't this whole function be `return not self.can_collect(context)`?

Reply:
I'll have to draw a table of results for each state and compare, because I'm not sure.